LayoutTests/accessibility/mac/dynamic-selection.html

```diff
@@ -0,0 +1,56 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
+<html>
+<head>
+<script src="../../resources/accessibility-helper.js"></script>
+<script src="../../resources/js-test.js"></script>
+</head>
+<body>
+
+Foo <span id="span-one">bar baz</span>
+One <span id="span-two">two three</span>
+
+<script>
+var output = "This test ensures we update the accessibility tree when selected text changes dynamically.\n\n";
+
+async function selectElementTextById(id) {
+    const webArea = accessibilityController.rootElement.childAtIndex(0);
+    const selection = window.getSelection();
+    selection.removeAllRanges();
+
+    await waitFor(() => {
+        const selectedRange = webArea.selectedTextMarkerRange();
+        return webArea.textMarkerRangeLength(selectedRange) == 0;
+    });
+
+    const range = document.createRange();
+    range.selectNodeContents(document.getElementById(id));
+    selection.addRange(range);
+
+    let selectedRange;
+    await waitFor(() => {
+        selectedRange = webArea.selectedTextMarkerRange();
+        return webArea.textMarkerRangeLength(selectedRange) > 0;
+    });
+    return selectedRange;
+}
+
+if (window.accessibilityController) {
+    window.jsTestIsAsync = true;
+
+    var webArea = accessibilityController.rootElement.childAtIndex(0);
+    var selectedRange;
+    setTimeout(async function() {
+        selectedRange = await selectElementTextById("span-one");
+        output += expect("webArea.stringForTextMarkerRange(selectedRange)", "'bar baz'");
+
+        selectedRange = await selectElementTextById("span-two");
+        output += expect("webArea.stringForTextMarkerRange(selectedRange)", "'two three'");
+
+        debug(output);
+        finishJSTest();
+    }, 0);
+}
+</script>
+</body>
+</html>
```

Comment:
This test is fine and it follows a pattern we have implemented in many tests. A more realistic test in the sense of mimicking AT interaction more closely, we should consider moving towards an event notification driven model. This is that instead of waitFor, use a listener for the selection changed notification and query for the new notification in the listener. That's how AT operates and the test will check both the correct notifications and the property retrieval. Not saying that this test is wrong, but perhaps this is a good turning point to further adopt the event driven model in our tests.

Reply:
Good idea, let's take that strategy for future tests.